src/bigtable/mod.rs

```diff
@@ -0,0 +1,594 @@
+            }
+        }
+    }
+
+    fn new_family(&mut self, name: String) {
+        self.flush_bytes_in_progress();
+        if self.row.families.last().map(|fam| &fam.name) != Some(&name) {
+            self.row.families.push(Family {
+                name,
+                ..Default::default()
+            });
+        }
+    }
+
+    fn new_column(&mut self, qualifier: Bytes) {
+        self.flush_bytes_in_progress();
+        if let Some(family) = self.row.families.last_mut() {
+            if family.columns.last().map(|col| &col.qualifier) != Some(&qualifier) {
+                family.columns.push(Column {
+                    qualifier,
+                    ..Default::default()
+                });
+            }
+        }
+    }
+
+    fn last_column_mut(&mut self) -> Option<&mut Column> {
+        self.row
+            .families
+            .last_mut()
+            .and_then(|family| family.columns.last_mut())
+    }
+
+    fn new_cell(&mut self, timestamp_micros: i64, value_size: i32) {
+        self.flush_bytes_in_progress();
+        if let Some(col) = self.last_column_mut() {
+            col.cells.push(Cell {
+                timestamp_micros,
+                ..Default::default()
+            });
+            self.value = BytesMut::with_capacity(value_size as usize);
+        }
+    }
+
+    fn last_cell_mut(&mut self) -> Option<&mut Cell> {
+        self.last_column_mut().and_then(|col| col.cells.last_mut())
+    }
+
+    fn finish_row(&mut self) -> Row {
+        self.flush_bytes_in_progress();
+        std::mem::replace(&mut self.row, Default::default())
+    }
+
+    // Process a chunk, and return a row if one was completed.
+    //
+    // Bigtable responds in chunks, where a row can be split across chunks (but every chunk
+    // contains at most one row).
+    fn process_chunk(&mut self, chunk: v2::read_rows_response::CellChunk) -> Option<Row> {
+        if !chunk.row_key.is_empty() {
+            // We don't need to check if there's an existing row to store, because
+            // RowStatus tells us when to do that.
+            self.row.key = chunk.row_key;
+        }
+
+        if let Some(family_name) = chunk.family_name {
+            self.new_family(family_name);
+        }
+
+        if let Some(qualifier) = chunk.qualifier {
+            self.new_column(qualifier.into());
+        }
+
+        if Some(chunk.timestamp_micros) != self.last_cell_mut().map(|cell| cell.timestamp_micros) {
+            self.new_cell(chunk.timestamp_micros, chunk.value_size);
+        }
+
+        // TODO: in the (presumably reasonably common case that the buffer comes
+        // in a single chunk, we should avoid copying the data.
+        self.value.extend_from_slice(&chunk.value);
+
+        if let Some(row_status) = chunk.row_status {
+            let row = self.finish_row();
+            if matches!(
+                row_status,
+                v2::read_rows_response::cell_chunk::RowStatus::CommitRow(_),
```

Comment:
There isn't a handler for the other RowStatus variant, `ResetRow` which requires the data for the row key to be discarded.

Given that functionality, it might be better to accumulate a `Vec<Bytes>` (or better yet `SmallVec<[Bytes; 4]>`) instead of copying bytes into a `BytesMut`. Then a reset doesn't incur wasted copying, and as a bonus the single-chunk case gets zero copies

Reply:
I think this is actually correct, although I agree that is isn't clear, and that `SmallVec<Bytes>` makes sense: `self.finish_row()` clears the current row, so if we get the `ResetRow`, we clear the current row from our state and return `None`.